node/actors/network/src/gossip/attestation/mod.rs

```diff
@@ -0,0 +1,330 @@
+        votes: impl Iterator<Item = Arc<attester::Signed<attester::Batch>>>,
+    ) -> anyhow::Result<()> {
+        let mut done = HashSet::new();
+        for vote in votes {
+            // Disallow multiple entries for the same key:
+            // it is important because a malicious attester may spam us with
+            // new versions and verifying signatures is expensive.
+            if done.contains(&vote.key) {
+                anyhow::bail!("duplicate entry for {:?}", vote.key);
+            }
+            done.insert(vote.key.clone());
+            self.insert_vote(vote)?;
+        }
+        Ok(())
+    }
+
+    fn qc(&self) -> Option<attester::BatchQC> {
+        if self.weight < self.config.committee.threshold() {
+            return None;
+        }
+        let mut sigs = attester::MultiSig::default();
+        for vote in self.votes.values() {
+            sigs.add(vote.key.clone(), vote.sig.clone());
+        }
+        Some(attester::BatchQC {
+            message: self.config.batch_to_attest.clone(),
+            signatures: sigs,
+        })
+    }
+}
+
+/// Receiver of state diffs.
+pub(crate) struct DiffReceiver {
+    prev: Option<State>,
+    recv: sync::watch::Receiver<Option<State>>,
+}
+
+impl DiffReceiver {
+    /// Waits for the next state diff.
+    pub(crate) async fn wait_for_diff(&mut self, ctx: &ctx::Ctx) -> ctx::OrCanceled<Diff> {
+        loop {
+            let Some(new) = (*sync::changed(ctx, &mut self.recv).await?).clone() else {
+                continue;
+            };
+            let diff = new.diff(&self.prev);
+            self.prev = Some(new);
+            if !diff.is_empty() {
+                return Ok(diff);
+            }
+        }
+    }
+}
+
+/// `Controller` manages the attestation state.
+/// It maintains a set of votes matching the attestation config.
+/// It allows for
+/// * adding votes to the state
+/// * subscribing to the vote set changes
+/// * waiting for the certificate to be collected
+///
+/// It also keeps an attester key used to sign the batch vote,
+/// whenever it belongs the current attester committee.
+/// Signing happens automatically whenever the committee is updated.
+///
+/// Expected usage:
+/// ```
+/// let ctrl = attestation::Controller::new(Some(key));
+/// loop {
+///     // Check the global attestation registry.
+///     // Compute the next expected batch and the committee that should attest it.
+///     ...
+///     let config = attestation::Config {
+///         batch_to_attest: ...,
+///         committee: ...,
+///     };
+///     ctrl.update_config(Arc::new(config.clone())).unwrap();
+///     s.spawn(async {
+///         if let Some(qc) = ctrl.wait_for_qc(ctx, config.batch_to_attest.number).await?;
+///         // Submit the certificate `qc` to the global registry
+///         ...
+///     });
+///     // Wait for the global registry to include the certificate.
+///     ...
+/// }
+/// ```
+pub struct Controller {
+    key: Option<attester::SecretKey>,
```

Comment:
please add a comment what `None` and `Some` means here

Reply:
done